workflow/sync/sync_manager.go

```diff
@@ -0,0 +1,360 @@
+package sync
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/argoproj/argo/errors"
+	wfv1 "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	"github.com/argoproj/argo/util/slice"
+)
+
+type ReleaseNotifyCallbackFunc func(string)
+type SyncLimitConfigFunc func(string) (int, error)
+
+type SyncManager struct {
+	syncLockMap         map[string]Synchronization
+	lock                *sync.Mutex
+	releaseNotifyFunc   ReleaseNotifyCallbackFunc
+	syncLimitConfigFunc SyncLimitConfigFunc
+}
+
+type LockName struct {
+	Namespace    string
+	Kind         string
+	ResourceName string
+	Key          string
+}
+
+type Synchronization interface {
+	acquire(holderKey string) bool
+	tryAcquire(holderKey string) (bool, string)
+	release(key string) bool
+	addToQueue(holderKey string, priority int32, creationTime time.Time)
+	getCurrentHolders() []string
+	getName() string
+	getLimit() int
+	resize(n int) bool
+}
+
+type LockAction string
+
+const (
+	acquired LockAction = "acquired"
```

Comment:
Should this be
```suggestion
	LockActionAcquired LockAction = "Acquired"
```
Or is there a reason you've named the variables that way?